New command pref attempt

coinzdude opened the pull request:
rough attempt to start user prefs outside of projectstate

Rowdster committed:
font tweaks for commands

Rowdster committed:
more command layout tweaks. These fix the runtime colors

Rowdster committed:
tighten command spacing

Rowdster committed:
tweak for PR

Rowdster committed:
Merge branch 'trunk' of https://github.com/SeleniumHQ/selenium-ide into trunk

Rowdster committed:
More space for commands by reducing appbar, hack to scroll playback selected command

Rowdster committed:
Merge branch 'trunk' of https://github.com/SeleniumHQ/selenium-ide into trunk

Rowdster committed:
Merge branch 'trunk' of https://github.com/SeleniumHQ/selenium-ide into trunk

Rowdster committed:
initial pass to get save confirmation on project load, app close - close not working, cancel not working

coinzdude committed:
Merge branch 'SeleniumHQ:trunk' into trunk

coinzdude committed:
Merge branch 'SeleniumHQ:trunk' into trunk

Rowdster committed:
attempt at command layout

Rowdster committed:
some cleanup

toddtarsi committed:
text overflow ellipsis and multi command generalization

Rowdster committed:
Initial support for persistent window sizing

Rowdster committed:
bring project window to front on start

Rowdster committed:
initial code to preserve project state on reload

Rowdster committed:
code builds, cannot load prior project state but default works

Rowdster committed:
updated to not use default, cannot load project

Rowdster committed:
save and reload project state works!

Rowdster committed:
fix for loading unitialized projects

Rowdster committed:
updated for PR feedback

Rowdster committed:
Experiment on doClick to considert fallbacks

Rowdster committed:
Fix for selectFrame matching issue

Rowdster committed:
remove this.driver.switchTo().defaultContent

Rowdster committed:
merge cleanup for selectFrame-fix

coinzdude committed:
Merge branch 'SeleniumHQ:trunk' into trunk

Rowdster committed:
rimraf

Rowdster committed:
VS Code LF mode

Rowdster committed:
try catch on bad input files

Rowdster committed:
fixes for dev cli inputs

Rowdster committed:
Merge branch 'handle-bad-files' into trunk

Rowdster committed:
Merge branch 'selectFrame-fix' into trunk

Rowdster committed:
Merge branch 'trunk' of https://github.com/SeleniumHQ/selenium-ide into trunk

coinzdude committed:
Merge branch 'SeleniumHQ:trunk' into trunk

Rowdster committed:
Merge branch 'trunk' of https://github.com/coinzdude/selenium-ide into trunk

Rowdster committed:
remove console log on winow size changes

coinzdude committed:
Merge branch 'SeleniumHQ:trunk' into trunk

Rowdster committed:
Merge branch 'trunk' of https://github.com/coinzdude/selenium-ide into trunk

Rowdster committed:
try to move user pref to storage

Rowdster committed:
Start on a UserPrefs model

Rowdster committed:
working on user prefs

toddtarsi commented on packages/selenium-ide/src/api/commands/state/toggleUserPref.ts:
```diff
@@ -0,0 +1,34 @@
+
+export const browser = browserHandler<Shape>()
+
+export const main = mainHandler<Shape>(passthrough)
```
@coinzdude - Remove passthrough from here. This tells the server basically to immediately mutate and emit the onMutate event. In this case, you want to do some storage stuff, so it isn't a passthrough

coinzdude replied:
I think I got this

coinzdude replied:
I had missed the unused passthrough which was caught on compile.  Fixed.

toddtarsi commented on packages/selenium-ide/src/api/commands/state/toggleUserPref.ts:
```diff
@@ -0,0 +1,34 @@
+
+export const main = mainHandler<Shape>(passthrough)
+
+export const setUserPrefs = (prefs: UserPrefs) => {
```
@coinzdude - This needs to not be here. Instead, make a method called toggleUserPref on StateController that does what you want.

coinzdude replied:
Moved

Rowdster committed:
more updates

toddtarsi commented on packages/selenium-ide/src/api/helpers/userPrefs.ts:
```diff
@@ -0,0 +1,15 @@
+import { UserPrefs } from 'api/types'
+import storage from 'main/store'
+
+export const getUserPrefs = (): UserPrefs => {
```
@coinzdude - This file needs to go

toddtarsi commented on packages/selenium-ide/src/browser/windows/ProjectEditor/tabs/Project/ProjectTab.tsx:
```diff
@@ -57,6 +61,24 @@ const ProjectTab: FC<ProjectTabProps> = ({ session: { project } }) => (
+            name="insertNewCommandPref"
+            value={'After'}
+            onChange={(e: any) => {
+              // toggleUserPref({insertCommandPref: e.target.value})
```
This should be `window.sideAPI.state.toggleUserPref`

coinzdude replied:
Ok, think I got this.

Rowdster committed:
flailing on the code now

Rowdster committed:
removed passthrough

Rowdster committed:
remove unused ref

Rowdster committed:
compiles and runs, toggle not yet working

Rowdster committed:
small but not yet working revisions

Rowdster committed:
ongoing tweaks

Rowdster committed:
merge conflict fix

coinzdude committed:
Merge branch 'trunk' into new-command-pref

toddtarsi commented on packages/selenium-ide/src/api/commands/state/toggleUserPref.ts:
```diff
@@ -0,0 +1,18 @@
+
+export type Shape = Session['state']['toggleUserPref'];
+
+export const mutator: Mutator<Shape> = (session) =>
```
export const mutator: Mutator<Shape> = (session, { params: [newPrefValue] }) =>

toddtarsi commented on packages/selenium-ide/src/api/commands/state/toggleUserPref.ts:
```diff
@@ -0,0 +1,18 @@
+export const mutator: Mutator<Shape> = (session) =>
+  update(
+    'state.insertCommandPref',
+    (pref) => (pref === 'Before' ? 'After' : 'Before'),
```
(pref) => newPrefValue,

coinzdude replied:
This one is confusing to me, but I did implement as an update.

toddtarsi commented on packages/selenium-ide/src/browser/windows/ProjectEditor/tabs/Project/ProjectTab.tsx:
```diff
@@ -57,6 +61,23 @@ const ProjectTab: FC<ProjectTabProps> = ({ session: { project } }) => (
+            name="insertNewCommandPref"
+            value={state.userPrefs.insertCommandPref}
+            onChange={(e: any) => {
+              window.sideAPI.state.toggleUserPref(e.target.value)
```
This is being called with only a string on the browser side, and then you've got a controller function on the other side with some object type. Your linter should not be okay with one side of this. It's a very clear disagreement of types.

coinzdude replied:
I'm not seeing any linter complaint, but it's also not yet working

Rowdster committed:
some updates attempting to work through the user preference setting

Rowdster committed:
some updates attempting to work through the user preference setting

toddtarsi commented on packages/selenium-ide/src/main/session/controllers/Windows/index.ts:
```diff
@@ -170,6 +170,10 @@ export default class WindowsController extends BaseController {
     const size = storage.get<'windowSize'>('windowSize')
     const position = storage.get<'windowPosition'>('windowPosition')
+
+    // const userPrefs = storage.get<'insertCommandPref'>('insertCommandPref')
```
@coinzdude - The changes in this file should be dropped.

toddtarsi commented on packages/selenium-ide/src/browser/windows/ProjectEditor/tabs/Project/ProjectTab.tsx:
```diff
@@ -57,6 +61,23 @@ const ProjectTab: FC<ProjectTabProps> = ({ session: { project } }) => (
+              window.sideAPI.state.toggleUserPrefInsert(e.target.value)
+            }}
+          >
+            <MenuItem value={'After'}>After</MenuItem>
```
You can simplify static strings like this in react. I'm not sure if they currently do but I think they can optimize for this case in some way.

```
<MenuItem value="After">
```

coinzdude replied:
Good to have, thanks

coinzdude replied:
I tried that and the drawn select list options were blank. I didn't find a better or more succinct way and only see others using some kind of onclick handler and then processing the value out of the passed event. That seems like a bit of code overhead to just get what was selected.  

Here's a mega thread with the bundles of options on writing click handlers. Some of the examples here are wow. 
 https://stackoverflow.com/questions/43870814/get-value-of-menuitem-material-ui

toddtarsi replied:
Do you mean the quoting style won't work? I know we use this component without issue here:

https://github.com/SeleniumHQ/selenium-ide/blob/eaf634a3728ed756344a8969b7cb0cdc5848f3a0/packages/selenium-ide/src/browser/windows/ProjectEditor/tabs/Tests/CommandFields/CommandSelector.tsx#L49

toddtarsi replied:
Eh, we do this quoting style elsewhere unfortunately. Likely this doesn't have anything to do with it. My bet would be that you need those mutator changes to properly update the session shape.

coinzdude replied:
I have this using the simpler syntax, and in debugging, I can see that the state.userPrefs.insertCommandPref is changing and being set, but the UI is not getting or using the newer updated state info. It's stuck, seemingly hardcoded to 'After'

toddtarsi commented on packages/selenium-ide/src/api/commands/state/toggleUserPrefInsert.ts:
```diff
@@ -0,0 +1,20 @@
+    'state.userprefs.insertCommandPref',
+    (userPrefs: UserPrefs) =>
+      userPrefs.insertCommandPref = insertCommandPref
+      ,
```
I believe lodash fp does "immutable" type operations. They are designed to not mutate the original shape but return a new shape. You know, so that way we can make obtuse reduce functions and brick things on shit like simple counters. In a case where you're just setting a nested key, lodash/fp/set is probably a better choice than lodash/fp/update:

```
import set from 'lodash/fp/set'

export const mutator: Mutator<Shape> = (session, { params: [insertCommandPref] }) => set(
  'state.userprefs.insertCommandPref',
  insertCommandPref,
  session,
);
```

Something like this should work ^^^

coinzdude replied:
almost following, ha!  thanks!

toddtarsi commented:
Okay, those are all of my critiques about the code for now. They're largely pedantic, and meant to be half helpful and half just purely anxious about code smell about little things and such. All and all, this feature is shaping up well. Nice work!
I actually really like your most recent commit messages in git. They're small and frustrated, but they remind me of things I type for commit messages when I'm in serious iteration mode. I do still think you want to get in the practice of having minimal commits in a PR. I think the "ideal" currently is a PR as 1 commit with basically a couple paragraphs talking about it.

Rowdster committed:
PR tweaks

coinzdude commented:
I made the change to use 'set' instead of change in the mutator, and it seems like the code is running which updates the state, but I don't yet see that the user selection impacts the prefs tab for the select list.

Rowdster committed:
fix projecttab

Rowdster committed:
fix projecttab

Rowdster committed:
insert preference and usage on editing

toddtarsi committed:
fixes to command

properly source state as a react prop
correct capitalization on lodash set command

Rowdster committed:
Consider InsertCommandPref when adding new commands

coinzdude commented:
Last bit of code submitted now to load InsertCommandPref and userPrefs on project load and to consider this value when adding commands
